### PR TITLE
#1673 - Re-render Clipping Corners when User Interface Style Changes

### PR DESCRIPTION
**Pull request by shamanskyh**
This PR addresses an issue I raised with #1673 where the clipping corner layer doesn't get re-rendered when going from a dark to light user interface style. The fix is to add code in the `-asyncTraitCollectionDidChangeWithPreviousTraitCollection:` method that calls through to `-_updateClipCornerLayerContentsWithRadius:backgroundColor:` if the user interface style changes.

I've also added snapshot tests to verify that using clipping corners with a dynamic background color updates when the user interface style changes.

### CODE REVIEW

File: Source/ASDisplayNode.mm [hunk 1]
```diff
@@ -442,16 +442,29 @@ - (void)asyncTraitCollectionDidChangeWithPreviousTraitCollection:(ASPrimitiveTra
...
       // This change doesn't happen frequently so it's fairly safe to render nodes again
-      if (_loaded(self) && self.isLayerBacked) {
-        // Background colors do not dynamically update for layer backed nodes since they utilize CGColorRef
-        // instead of UIColor. We utilize the _backgroundColor instance variable to track the full dynamic color
-        // and apply any changes here when trait collection updates occur.
-        CGColorRef cgBackgroundColor = _backgroundColor.CGColor;
-        if (!CGColorEqualToColor(_layer.backgroundColor, cgBackgroundColor)) {
-          _layer.backgroundColor = cgBackgroundColor;
+      BOOL needsClippingCornerUpdate = NO;
+      CGFloat cornerRadius = _cornerRadius;
```
Comment (shamanskyh) [inline Source/ASDisplayNode.mm]: Capturing these values up here so they're behind the lock, but in scope when we call `_updateClipCornerLayerContentsWithRadius` on line 466

File: Tests/ASDisplayNodeSnapshotTests.mm [hunk 2]
```diff
@@ -61,18 +61,50 @@ - (void)testPrecompositedCornerRounding
 
 - (void)testClippingCornerRounding
 {
+  self.recordMode = YES;
+#if AS_AT_LEAST_IOS13
```
Comment (shamanskyh) [inline Tests/ASDisplayNodeSnapshotTests.mm]: Looking for feedback on this inline approach. Should I delete the reference images for older Xcode/iOS versions?

Comment (bolsinga) [reply Tests/ASDisplayNodeSnapshotTests.mm]: I personally think this is fine for now.

File: Source/ASDisplayNode.mm [hunk 3]
```diff
@@ -442,16 +442,29 @@ - (void)asyncTraitCollectionDidChangeWithPreviousTraitCollection:(ASPrimitiveTra
...
+      ASCornerRoundingType cornerRoundingType = _cornerRoundingType;
+      UIColor *backgroundColor = self.backgroundColor;
```
Comment (rahul-malik) [inline Source/ASDisplayNode.mm]: any reason we're using `self.backgroundColor` vs `_backgroundColor`

File: Source/ASDisplayNode.mm [hunk 4]
```diff
@@ -442,16 +442,29 @@ - (void)asyncTraitCollectionDidChangeWithPreviousTraitCollection:(ASPrimitiveTra
...
+      if (_loaded(self)) {
+        if (self.isLayerBacked) {
+          // Background colors do not dynamically update for layer backed nodes since they utilize CGColorRef
+          // instead of UIColor. We utilize the _backgroundColor instance variable to track the full dynamic color
+          // and apply any changes here when trait collection updates occur.
+          CGColorRef cgBackgroundColor = _backgroundColor.CGColor;
```
Comment (rahul-malik) [inline Source/ASDisplayNode.mm]: use should use the new local variable here

File: Tests/ASDisplayNodeSnapshotTests.mm [hunk 5]
```diff
@@ -61,18 +61,49 @@ - (void)testPrecompositedCornerRounding
...
+    }
+#else
     node.backgroundColor = UIColor.greenColor;
+#endif
     node.maskedCorners = c;
     node.cornerRadius = 15;
     // A layout pass is required, because that's where we lay out the clip layers.
     [node.layer layoutIfNeeded];
+
+#if AS_AT_LEAST_IOS13
```
Comment (rahul-malik) [inline Tests/ASDisplayNodeSnapshotTests.mm]: We should probably write a macro for light/dark testing if we're doing this a lot

File: Source/ASDisplayNode.mm (hunk 3 above)
Comment (shamanskyh) [reply Source/ASDisplayNode.mm]: `self.backgroundColor` has a getter that pulls from the layer or the view. `_backgroundColor` should only be used if layer backed.

File: Source/ASDisplayNode.mm (hunk 4 above)
Comment (shamanskyh) [reply Source/ASDisplayNode.mm]: Done

**Pull request merged**